web/api/v1/api.go

```diff
@@ -464,6 +475,82 @@ func (api *API) queryRange(r *http.Request) (result apiFuncResult) {
 	}, nil, res.Warnings, qry.Close}
 }
 
+func (api *API) queryExemplars(r *http.Request) apiFuncResult {
+	start, err := parseTime(r.FormValue("start"))
+	if err != nil {
+		err = errors.Wrapf(err, "invalid parameter 'start'")
+		return apiFuncResult{nil, &apiError{errorBadData, err}, nil, nil}
+	}
+	end, err := parseTime(r.FormValue("end"))
+	if err != nil {
+		err = errors.Wrapf(err, "invalid parameter 'end'")
+		return apiFuncResult{nil, &apiError{errorBadData, err}, nil, nil}
+	}
+	if end.Before(start) {
+		err := errors.New("end timestamp must not be before start time")
+		return apiFuncResult{nil, &apiError{errorBadData, err}, nil, nil}
+	}
+
+	ctx := r.Context()
+	if to := r.FormValue("timeout"); to != "" {
+		var cancel context.CancelFunc
+		timeout, err := parseDuration(to)
+		if err != nil {
+			err = errors.Wrap(err, "invalid parameter 'timeout'")
+			return apiFuncResult{nil, &apiError{errorBadData, err}, nil, nil}
+		}
+
+		ctx, cancel = context.WithTimeout(ctx, timeout)
+		defer cancel()
+	}
+
+	q, err := api.Queryable.Querier(ctx, timestamp.FromTime(minTime), timestamp.FromTime(maxTime))
+	if err != nil {
+		return apiFuncResult{nil, &apiError{errorBadData, err}, nil, nil}
+	}
+	defer q.Close()
+
+	expr, err := parser.ParseExpr(r.FormValue("query"))
+	if err != nil {
+		return apiFuncResult{nil, &apiError{errorBadData, err}, nil, nil}
+	}
+
+	selectors, err := parser.ExtractSelectors(expr)
+	if err != nil {
+		return apiFuncResult{nil, &apiError{errorBadData, err}, nil, nil}
+	}
+	if len(selectors) < 1 {
+		return apiFuncResult{nil, &apiError{errorBadData, fmt.Errorf("no series found for supplied query: %s", expr)}, nil, nil}
+	}
+
+	var retExemplars []exemplarData
+	var series []storage.SeriesSet
+
+	for _, selectorArr := range selectors {
+		series = append(series, q.Select(false, nil, selectorArr...))
+	}
+
+	for _, s := range series {
+		eq, err := api.ExemplarQueryable.Querier(context.Background())
+		if err != nil {
+			return apiFuncResult{nil, &apiError{errorBadData, err}, nil, nil}
+		}
+
+		for s.Next() {
+			res, err := eq.Select(&storage.SelectHints{Start: timestamp.FromTime(start), End: timestamp.FromTime(end)}, s.At().Labels())
```

Comment:
Why are we passing storage hints for times, surely they should be first class parameters?

Reply:
I based this on the fact that the regular queriers select function takes `SelectHints`, but `NewQuerier` takes the start and end as first class parameters. I prefer the first class approach, I just wasn't sure how to take the "exemplar storage and interfaces should align with the existing interfaces" approach.

Reply:
Hints are just hints, the correct result should still be provided without them.